src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes managed and unmanaged resources held by this instance.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> if <see cref="Dispose()"/> was called; <c>false</c> if the object is being finalized.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.semaphore.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Throws an exception if this instance has been faulted.
+        /// </summary>
+        protected virtual void ThrowIfFaulted()
+        {
+        }
+
+        /// <summary>
+        /// Disposes the specfied release, swallowing certain exceptions.
+        /// </summary>
+        /// <param name="releaser">The releaser to dispose.</param>
+        private static void DisposeReleaserNoThrow(AsyncSemaphore.Releaser releaser)
+        {
+            try
+            {
+                releaser.Dispose();
+            }
+            catch (ObjectDisposedException)
+            {
+                // Swallow this, since in releasing the semaphore if it's already disposed the caller probably doesn't care.
+            }
+        }
+
+        /// <summary>
+        /// An implementation of <see cref="ReentrantSemaphore"/> supporting the <see cref="ReentrancyMode.NotRecognized"/> mode.
+        /// </summary>
+        private class NotRecognizedSemaphore : ReentrantSemaphore
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="NotRecognizedSemaphore"/> class.
+            /// </summary>
+            /// <param name="initialCount">The initial number of concurrent operations to allow.</param>
+            /// <param name="joinableTaskContext">The <see cref="JoinableTaskContext"/> to use to mitigate deadlocks.</param>
+            internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTaskContext)
+                : base(initialCount, joinableTaskContext)
+            {
+            }
+
+            /// <inheritdoc />
+            public override async Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default)
+            {
+                Requires.NotNull(operation, nameof(operation));
+                this.ThrowIfFaulted();
+
+                Func<Task> semaphoreUser = async delegate
+                {
+                    using (this.joinableTaskCollection?.Join())
+                    {
+                        var releaser = await this.semaphore.EnterAsync(cancellationToken).ConfigureAwait(true);
+                        try
+                        {
+                            await operation().ConfigureAwaitRunInline();
```

Comment:
>operation() [](start = 34, length = 11)

Might a caller also want to have ```cancellationToken``` passed into ```operation```?  Is it worth another overload of ```ExecuteAsync```?

Reply:
The operation may need any number of inputs, that would be implicitly captured in its closure, including `cancellationToken`. If we choose to pass anything into `operation`, I think I'd make it a `T` or object state value with the goal of allowing the caller to avoid the cost of the closure and delegate allocations by storing N things in the object that they want me to propagate. Propagating just one value IMO just requires the caller to give a second name to a CancellationToken they already have access to.